File: skills/math-expression-parser/reference/src/evaluator.ts
/**
 * Evaluates an AST node to produce a numeric result.
 *
 * Supports: +, -, *, /, %, ** (power), unary negation.
 * Throws on division by zero and modulo by zero.
 *
 * @node evaluator
 * @depends-on ast-types
 * @contract evaluator.test.ts
 * @hint evaluator: Simple recursive tree walk. Match on node type,
 *       recurse into children, apply the operator.
 */

import { type AstNode } from "./ast-types";

export function evaluate(node: AstNode): number {
  if (node.type === "number") {
    return node.value;
  }

  if (node.type === "unary") {
    return -evaluate(node.operand);
  }

  const left = evaluate(node.left);
  const right = evaluate(node.right);

  if (node.op === "+") return left + right;
  if (node.op === "-") return left - right;
  if (node.op === "*") return left * right;
  if (node.op === "**") return left ** right;

  if (node.op === "/") {
    if (right === 0) throw new Error("Division by zero");
    return left / right;
  }

  // node.op === "%"
  if (right === 0) throw new Error("Modulo by zero");
  return left % right;
}

File: research/experiments/sweep/runs/mathexpr-full-cpp-sonnet45/src/evaluator.cpp
#include "evaluator.h"
#include <stdexcept>
#include <cmath>

double evaluate(std::shared_ptr<AstNode> node) {
    if (node->type == AstNodeType::Number) {
        auto numNode = std::dynamic_pointer_cast<NumberLiteral>(node);
        return numNode->value;
    }

    if (node->type == AstNodeType::Unary) {
        auto unaryNode = std::dynamic_pointer_cast<UnaryExpr>(node);
        double operandValue = evaluate(unaryNode->operand);
        if (unaryNode->op == "-") {
            return -operandValue;
        }
        throw std::runtime_error("Unknown unary operator");
    }

    if (node->type == AstNodeType::Binary) {
        auto binNode = std::dynamic_pointer_cast<BinaryExpr>(node);
        double leftValue = evaluate(binNode->left);
        double rightValue = evaluate(binNode->right);

        if (binNode->op == "+") {
            return leftValue + rightValue;
        } else if (binNode->op == "-") {
            return leftValue - rightValue;
        } else if (binNode->op == "*") {
            return leftValue * rightValue;
        } else if (binNode->op == "/") {
            if (rightValue == 0) {
                throw std::runtime_error("Division by zero");
            }
            return leftValue / rightValue;
        } else if (binNode->op == "%") {
            if (rightValue == 0) {
                throw std::runtime_error("Modulo by zero");
            }
            return std::fmod(leftValue, rightValue);
        } else if (binNode->op == "**") {
            return std::pow(leftValue, rightValue);
        }
        throw std::runtime_error("Unknown binary operator");
    }

    throw std::runtime_error("Unknown node type");
}
